src/main/java/uk/gov/ons/ctp/response/action/service/ActionProcessingService.java

```diff
@@ -35,66 +33,119 @@
 
-  public ActionProcessingService(ActionRequestDecorator[] decorators) {
-    this.decorators = decorators;
+  public void processActions(ActionType actionType, List<Action> allActions) {
+    if (actionType == null) {
+      throw new IllegalStateException(ACTION_TYPE_NOT_DEFINED);
+    }
+    String handler = actionType.getHandler();
+    if (handler == null) {
+      log.with("handler", actionType.getHandler()).error("no action type handler provided");
+    } else if (handler.equals(NOTIFY)) {
+      processEmails(actionType, allActions);
+    } else if (actionType.getHandler().equals(PRINTER)) {
+      processLetters(actionType, allActions);
+    } else {
+      log.with("handler", actionType.getHandler())
+          .with("actions", allActions.size())
+          .error("unsupported action type handler");
+    }
   }
 
-  /** Distributes requests for a single action */
-  @Transactional(propagation = Propagation.REQUIRES_NEW)
-  public void processActionRequests(final UUID actionId) {
-    Action action = actionRepo.findById(actionId);
-    log.with("actionId", action.getId()).info("Processing actionRequest");
+  public void processLetters(ActionType actionType, List<Action> allActions) {
+    // action requests are decorated actions
+    log.with("actions", allActions.size()).debug("processing letters");
+    List<ActionRequest> printerActions = new ArrayList<>();
+    for (Action action : allActions) {
+      if (checkCaseExists(action) && validate(action) && !cancelled(action)) {
+        List<ActionRequest> actionRequests = prepareActionRequests(action);
+        printerActions.addAll(actionRequests);
+      }
+    }
+    log.with("entries", printerActions.size()).debug("about to create print files");
+    notificationFileCreator.export(getActionEvent(actionType), printerActions);
+  }
 
+  private boolean validate(final Action action) {
     final ActionType actionType = action.getActionType();
     if (!valid(actionType)) {
       log.with("action", action).error("ActionType is not defined for action");
-      throw new IllegalStateException(ACTION_TYPE_NOT_DEFINED);
+      return false;
+    } else {
+      return true;
     }
+  }
 
-    List<ActionRequest> actionRequests = prepareActionRequests(action);
-
-    ActionDTO.ActionEvent event =
-        actionType.getResponseRequired()
-            ? ActionDTO.ActionEvent.REQUEST_DISTRIBUTED
-            : ActionDTO.ActionEvent.REQUEST_COMPLETED;
-    transitionAction(action, event);
-
-    actionRequests.forEach(
-        actionRequest -> {
-          if (actionRequest.isPubsub()) {
-            log.with("actionId", actionId.toString())
-                .info("Pubsub message will be forwarded to notify cloudfunction");
+  public void processEmails(ActionType actionType, List<Action> allActions) {
+    log.with("actions", allActions.size()).debug("processing emails");
+    for (Action action : allActions) {
+      if (checkCaseExists(action) && validate(action) && !cancelled(action)) {
```

Comment:
Do we need to log anything if any of these fail?  Even just a simple `"Not going to process this"`

Reply:
Done 👍